## 1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.
## Importing Libraries

In [5]:
# Import libraries 
import pandas as pd
import numpy as np
import os

## Importing Data

In [6]:
# Turn project folder path into a string 
path = r'/Users/jerriejiang/Desktop/4.0/A4_Instacart Basket Analysis/02 Data/Original Data/'

In [7]:
path = r'/Users/jerriejiang/Desktop/4.0/A4_Instacart Basket Analysis'

In [10]:
# Import and define dataframe including newly derived columns from the previous exercise
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_new_variables_derived.pkl'))

In [8]:
# Check dimensions of imported dataframe for consistency
ords_prods_merge.shape

(32399732, 21)

## 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [4]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.667853
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

## 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

## Comparison with Subset Results:
The results from the entire dataframe include all 'department_id' values.
The mean values for the 'department_id's present in the subset have changed when compared to the entire dataframe.
Changes in Mean Values:
Some 'department_id' mean values have increased.
Some 'department_id' mean values have decreased.
Example:
In the subset, 'department_id' 4 had a mean value of 18.825780.
In the entire dataframe, 'department_id' 4 has a mean value of 17.811403.

## 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [15]:
# These operations were already performed on the dataframe during the exercise.

In [6]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   
15                     6.0        False         196                  3   
16                     8.0        False         196                 15   
17                     6.0        False         196                  1   
18                     6.0        False         196                  2   
19                     7.0        False         196                 17   

    reordered  ... aisle_id  department_id  prices  _merge    price_range_loc  \
0           0  ...       77              7     9.0    both  Mid-range product   
1           1  ...       77              7     9.0    both  Mid-range product   
2           1  ...       77              7     9.0    both  Mid-range product   
3           1  ...       77              7     9.0    both  Mid-range product   
4           1  ...       77              7     9.0    both  Mid-range product   
5           1  ...       77              7     9.0    both  Mid-range product   
6           1  ...       77              7     9.0    both  Mid-range product   
7           1  ...       77              7     9.0    both  Mid-range product   
8           1  ...       77              7     9.0    both  Mid-range product   


In [16]:
# The 'max_orders' and 'loyalty_flag' columns are already in place.

In [17]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

## 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [18]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'std', 'median']})

prices                                 
                       mean  min      max         std median
loyalty_flag                                                
Loyal customer    10.386336  1.0  99999.0  328.017787    7.4
New customer      13.294670  1.0  99999.0  597.560299    7.4
Regular customer  12.495717  1.0  99999.0  539.720919    7.4

#### Removing the outliers priced 99999 have strong influenced the 'mean'.
#### These means are different than other submissions. 

## 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”.
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender”.

In [19]:
# Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [20]:
# Check the new column
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    Unnamed: 0  order_id  user_id eval_set  order_number  ...  \
0         1987   3139998      138    prior            28  ...   
1         1989   1977647      138    prior            30  ...   
2        11433    389851      709    prior             2  ...   
3        12198    652770      764    prior             1  ...   
4        12200   1813452      764    prior             3  ...   
5        12372   1701441      777    prior            16  ...   
6        13096   1871483      825    prior             3  ...   
7        14603   1290456      910    prior            12  ...   
8        17065    369558     1052    prior            10  ...   
9        17070    589712     1052    prior            15  ...   
10       22082   1798115     1379    prior             1  ...   
11       23713    497193     1480    prior            36  ...   
12       23715   1727826     1480    prior            38  ...   
13       23716   2694698     1480    prior            39  ...   
14       23937   3171753     1494    prior             6  ...   
15       23938    186761     1494    prior             7  ...   
16       23940    317121     1494    prior             9  ...   
17       24740    200390     1540    prior             4  ...   
18       24741    374436     1540    prior             5  ...   
19       24742   2920553     1540    prior             6  ...   

    days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                      3.0                  5          0    both   
1                     20.0                  1          1    both   
2                      6.0                 20          0    both   
3                      NaN                 10          0    both   
4                      9.0                 11          1    both   
5                     26.0                  7          0    both   
6                     30.0                  2          0    both   
7                     30.0                  1          0    both   
8                     19.0                  1          0    both   
9                     15.0                  2          1    both   
10                     NaN                  1          0    both   
11                     7.0                  1          0    both   
12                     7.0                  3          1    both   
13                    12.0

In [21]:
# Assign spender tag to each customer
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [22]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [23]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

In [24]:
# The results show a relatively low number of high spenders, once again, likely due to the removal of outliers

## 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a 'Non-frequent customer'.
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a 'Regular customer'.
3. If the median is lower than or equal to 10, then the customer should be labeled a 'Frequent customer'.

In [25]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [26]:
# Check that the column was created successfully and properly
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    Unnamed: 0  order_id  user_id eval_set  order_number  ...  reordered  \
0         1987   3139998      138    prior            28  ...          0   
1         1989   1977647      138    prior            30  ...          1   
2        11433    389851      709    prior             2  ...          0   
3        12198    652770      764    prior             1  ...          0   
4        12200   1813452      764    prior             3  ...          1   
5        12372   1701441      777    prior            16  ...          0   
6        13096   1871483      825    prior             3  ...          0   
7        14603   1290456      910    prior            12  ...          0   
8        17065    369558     1052    prior            10  ...          0   
9        17070    589712     1052    prior            15  ...          1   
10       22082   1798115     1379    prior             1  ...          0   
11       23713    497193     1480    prior            36  ...          0   
12       23715   1727826     1480    prior            38  ...          1   
13       23716   2694698     1480    prior            39  ...          1   
14       23937   3171753     1494    prior             6  ...          0   
15       23938    186761     1494    prior             7  ...          1   
16       23940    317121     1494    prior             9  ...          1   
17       24740    200390     1540    prior             4  ...          0   
18       24741    374436     1540    prior             5  ...          1   
19       24742   2920553     1540    prior             6  ...          1   

    _merge    price_range_loc     busiest_day    busiest_days max_order  \
0     both  Mid-range product  Regularly busy  Regularly busy        32   
1     both  Mid-range product  Regularly busy  Regularly busy        32   
2     both  Mid-range product     Busiest day    Busiest days         5   
3     both  Mid-range product  Regularly busy    Slowest days         3   
4     both  Mid-range product      Least busy    Slowest days         3   
5     both  Mid-range product  Regularly busy    Busiest days        26   
6     both  Mid-range product  Regularly busy  Regularly busy         9   
7     both  Mid-range product  Regularly busy    Slowest days        12   
8     both  Mid-range product  Regularly busy    Busiest days        20 

In [27]:
# Based on the criteria provided, create a column featuring order frequency flags
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [28]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'

In [29]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [30]:
# Check value counts for new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [31]:
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(60)

user_id  customer_frequency         frequency_flag
0       138                 8.0      Frequent customer
1       138                 8.0      Frequent customer
2       709                 8.0      Frequent customer
3       764                 9.0      Frequent customer
4       764                 9.0      Frequent customer
5       777                11.0       Regular customer
6       825                20.0       Regular customer
7       910                 6.0      Frequent customer
8      1052                10.0      Frequent customer
9      1052                10.0      Frequent customer
10     1379                20.0       Regular customer
11     1480                 7.0      Frequent customer
12     1480                 7.0      Frequent customer
13     1480                 7.0      Frequent customer
14     1494                 7.0      Frequent customer
15     1494                 7.0      Frequent customer
16     1494                 7.0      Frequent customer
17     1540                 7.0      Frequent customer
18     1540                 7.0      Frequent customer
19     1540                 7.0      Frequent customer
20     1540                 7.0      Frequent customer
21     1540                 7.0      Frequent customer
22     1540                 7.0      Frequent customer
23     1540                 7.0      Frequent customer
24     1540                 7.0      Frequent customer
25     1540                 7.0      Frequent customer
26     1540                 7.0      Frequent customer
27     1540                 7.0      Frequent customer
28     1540                 7.0      Frequent customer
29     1540                 7.0      Frequent customer
30     1540                 7.0      Frequent customer
31     1540                 7.0      Frequent customer
32     1540                 7.0      Frequent customer
33     1540                 7.0      Frequent customer
34     1598                 2.0      Frequent customer
35     1598                 2.0      Frequent customer
36     1598                 2.0      Frequent customer
37     1647                25.0  Non-frequent customer
38     1647                25.0  Non-frequent customer
39     1647                25.0  Non-frequent customer
40     1647                25.0  Non-frequent customer
41     1647                25.0  Non-frequent customer
42     2833                13.0       Regular customer
43     2850                 7.0      Frequent customer
44     2857                11.0       Regular customer
45     3010                 7.0      Frequent customer
46     3029                28.0  Non-frequent customer
47     3395                18.0       Regular customer
48     3904                21.0  Non-frequent customer
49     4006                 7.0      Frequent customer
50     4006                 7.0      Frequent customer
51     4006                 7.0      Frequent customer
52     4006                 7.0      Frequent customer
53     4006                 7.0      Frequent customer
54     4122                 7.0      Frequent customer
55     4141                30.0  Non-frequent customer
56     4343                30.0  Non-frequent customer
57     4553                30.0  Non-frequent customer
58     4845                20.0       Regular customer
59     6866                30.0  Non-frequent customer

### 9. Export your dataframe as a pickle file and store it correctly in your 'Prepared Data' folder

In [32]:
# Check shape before exporting
ords_prods_merge.shape

(32404859, 25)

In [ ]:
# Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_merged_new_variables_derived_aggregated.pkl'))